fcrepo/fcrepo
# refactor tx to have state

**JIRA Ticket**: https://fedora-repository.atlassian.net/browse/FCREPO-3697

# What does this Pull Request do?

1. Adds a state to transactions
2. Ensures that changes cannot be made to a tx when it is no longer open
3. Waits for in progress changes to finish before committing a tx
4. If a request that's part of a long running tx fails, then the entire tx is marked as failed and cannot be committed

# How should this be tested?

1. Create a tx, create resources in the tx, add a resource that fails (such as an ag child of an ag), and see that the tx cannot be committed
2. Create a tx, concurrently add a large resource and a small resource in the tx, commit the tx before the large resource is finished, and see that the commit waits for it to finish before committing

# Interested parties
@fcrepo/committers
